File: location-java/src/main/java/com/location/LocationItem.java
package com.location;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public class LocationItem {
    private static final DateTimeFormatter DATE_FORMATTER =
            DateTimeFormatter.ofPattern("dd/MM/yyyy");

    private final String nom;
    private boolean estReserve;
    private LocalDate dateFinReservation;

    public LocationItem(String nom) {
        this.nom = nom;
        this.estReserve = false;
    }

    public String getNom() {
        return nom;
    }

    public boolean estDisponible() {
        if (!estReserve) {
            return true;
        }
        LocalDate aujourdHui = LocalDate.now();
        if (aujourdHui.isAfter(dateFinReservation)) {
            estReserve = false;
            dateFinReservation = null;
            return true;
        }
        return false;
    }

    public String reserver(int jours) {
        if (jours < 1) {
            return "La réservation doit durer au moins 1 jour.";
        }
        if (!estDisponible()) {
            return "L'objet '" + nom + "' est déjà réservé jusqu’au "
                    + dateFinReservation.format(DATE_FORMATTER) + ".";
        }
        estReserve = true;
        dateFinReservation = LocalDate.now().plusDays(jours);
        return "Réservation de '" + nom + "' pour " + jours + " jour(s).";
    }

    public String getStatut() {
        estDisponible();
        if (!estReserve) {
            return "Disponible";
        }
        return "Réservé jusqu’au " + dateFinReservation.format(DATE_FORMATTER);
    }
}

File: src/models/LocationItem.ts
import { format } from 'date-fns'

export class LocationItem {
  private nom: string
  private estReserve: boolean = false
  private dateFinReservation: Date | null = null

  constructor(nom: string) {
    this.nom = nom
  }

  public getNom(): string {
    return this.nom
  }

  public estDisponible(): boolean {
    if (!this.estReserve) return true

    const maintenant = new Date()
    if (this.dateFinReservation && maintenant > this.dateFinReservation) {
      this.estReserve = false
      this.dateFinReservation = null
      return true
    }

    return false
  }

  public reserver(jours: number): string {
    if (jours < 1) {
      return "La réservation doit durer au moins 1 jour."
    }

    if (!this.estDisponible()) {
      return `L'objet '${this.nom}' est déjà réservé jusqu’au ${format(this.dateFinReservation!, 'dd/MM/yyyy')}.`
    }

    this.estReserve = true
    const now = new Date()
    this.dateFinReservation = new Date(now.getTime() + jours * 24 * 60 * 60 * 1000)

    return `✅ Réservation de '${this.nom}' pour ${jours} jour(s).`
  }

  public getStatut(): string {
    this.estDisponible()
    if (!this.estReserve) {
      return "Disponible"
    }

    return `Réservé jusqu’au ${format(this.dateFinReservation!, 'dd/MM/yyyy')}`
  }
}
